test/jdk/java/util/stream/BuiltInGatherersTest.java

```diff
@@ -245,34 +253,43 @@ public void testMapConcurrentAPIandContract(Config config) throws InterruptedExc
         }
 
         // Test cancellation after exception during processing
-        if (config.streamSize > 2) { // We need streams of a minimum size to test this
-            final var firstLatch = new CountDownLatch(1);
-            final var secondLatch = new CountDownLatch(1);
-            final var cancellationLatch = new CountDownLatch(config.streamSize - 2); // all but two will get cancelled
+        // Only use reasonably sized streams to avoid excessive thread creation
+        if (config.streamSize > 2 && config.streamSize < 100) {
+            final var tasksToCancel = config.streamSize - 2;
+            final var throwerReady = new CountDownLatch(1);
+            final var initiateThrow = new CountDownLatch(1);
+            final var tasksCancelled = new CountDownLatch(tasksToCancel);
+
+            final var tasksWaiting = new Semaphore(0);
 
             try {
                 config.stream()
                       .gather(
                             Gatherers.mapConcurrent(config.streamSize(), i -> {
                                 switch (i) {
                                     case 1 -> {
-                                        try {
-                                            firstLatch.await(); // the first waits for the last element to start
-                                        } catch (InterruptedException ie) {
-                                            throw new IllegalStateException(ie);
-                                        }
+                                        throwerReady.countDown();
+                                        awaitSensibly(initiateThrow);
```

Comment:
20s might not be enough to wait for the count down when running with a debug build, -Xcomp, ...  I would be tempted to just have it wait indefinitely. If there is an issue then jtreg will timeout the test.